src/magic.rs

```diff
@@ -41,43 +41,39 @@ struct MagicRule {
 
 impl MagicRule {
     fn matches_data(&self, data: &[u8]) -> bool {
+        // Do we need the value_length at all, if it's implicit in
+        // value.len() and checked by the parsers?
+        assert!(self.value_length as usize == self.value.len());
+        assert!(self.mask.is_none() || self.mask.as_ref().unwrap().len() == self.value.len());
+
         let start: usize = self.start_offset as usize;
         let end: usize = self.start_offset as usize + self.range_length as usize;
-        let data_len: usize = data.len() as usize;
-
-        for i in start..end {
-            let mut res: bool = true;
 
-            let value_len: usize = self.value_length as usize;
+        let value_len = self.value.len();
 
-            if i + value_len > data_len {
+        for i in start..end {
+            if i + value_len > data.len() {
                 return false;
             }
 
+            let looking_at = &data[i..i + value_len];
+
             match &self.mask {
                 Some(m) => {
-                    for j in 0..value_len {
-                        let masked_value = self.value[j] & m[j];
-                        let masked_data = data[j + i] & m[j];
-                        if masked_value != masked_data {
-                            res = false;
-                            break;
+                    if looking_at
+                        .iter()
+                        .zip(m.iter())
+                        .map(|(l, m)| *l & *m)
+                        .eq(self.value.iter().map(|v| *v)) {
```

Comment:
This is kind of black magic and a bit hard to read for me.

The logic is that both the value in the rule and the data we're matching must be masked, but the `map()` seems to be masking only the `looking_at` vector.

Reply:
Oh, I see, indeed the original code did what you say :)  I've updated it with a more legible version.

Reply:
Also I updated the test for masks to actually need masking both the value and the data.